### PR TITLE
feat(esl-carousel): rework carousel plugins API to use json attr + smart media query

### PR DESCRIPTION
**Pull request by ala-n**

### CODE REVIEW

File: src/modules/esl-media-query/core/esl-media-rule-list.ts [hunk 1]
```diff
@@ -48,37 +48,46 @@ export class ESLMediaRuleList<T = any> extends SyntheticEventTarget {
...
    *
-   * @param values - values tuple string (uses '|' as separator)
-   * @param mask - media conditions tuple string (uses '|' as separator)
+   * @param query - media rule query ('arrow' syntax) string or tuple string of values (uses '|' as separator)
+   * @param mask - media conditions tuple string (uses '|' as separator), to be used in case of tuple syntax
    *
    * @example
    * ```ts
-   * ESLMediaRuleList.parse('@XS|@SM|@MD|@LG|@XL', '1|2|3|4|5')
+   * ESLMediaRuleList.parse('1 | @XS => 2', '1|2|3|4|5') // all => 1 | @xs => 2 (second argument is ignored)
```
Comment (NastaLeo) [inline src/modules/esl-media-query/core/esl-media-rule-list.ts:64]: please, pay attention that  in all examples the second argument should be mask

**Review by NastaLeo — APPROVED** [review]

File: src/modules/esl-carousel/plugin/esl-carousel.plugin.ts [hunk 2]
```diff
@@ -1,12 +1,52 @@
...
+  }
+
+  /**
+   * Parses plugin media query value term to the config object.
+   * Provides the capability to pass a config a stringified non-strict JSON or as a string (mapped to single option configuration).
+   *
+   * Uses {@link ESLCarouselPlugin.SHORT_OPTION} to map string value to the config object.
+   */
+  @bind
+  protected parseConfig(value: string): Config | null {
```
Comment (yadamskaya) [inline src/modules/esl-carousel/plugin/esl-carousel.plugin.ts:43]: AS I see from the return value it cannot be null
```suggestion
  protected parseConfig(value: string): Config {
```

**Review by abarmina — APPROVED** [review]

File: src/modules/esl-carousel/plugin/esl-carousel.plugin.ts (hunk 2 above)
Comment (ala-n) [reply src/modules/esl-carousel/plugin/esl-carousel.plugin.ts:43]: It could, line 44 )

Comment (yadamskaya) [reply src/modules/esl-carousel/plugin/esl-carousel.plugin.ts:43]: missed it :)

**Review by yadamskaya — APPROVED** [review]

File: src/modules/esl-carousel/plugin/esl-carousel.plugin.ts (hunk 2 above)
Comment (yadamskaya) [reply src/modules/esl-carousel/plugin/esl-carousel.plugin.ts:43]: missed it:)

File: src/modules/esl-carousel/plugin/esl-carousel.plugin.ts [hunk 3]
```diff
@@ -1,12 +1,52 @@
...
+import {bind, ready, memoize} from '../../esl-utils/decorators';
+import {evaluate} from '../../esl-utils/misc/format';
+import {ESLMediaRuleList} from '../../esl-media-query/core';
 import {ESLCarousel} from '../core/esl-carousel';
-import {ready} from '../../esl-utils/decorators/ready';
 
 /** Base mixin plugin of {@link ESLCarousel} */
-export abstract class ESLCarouselPlugin extends ESLMixinElement {
+export abstract class ESLCarouselPlugin<Config> extends ESLMixinElement {
+  /** Option to be aped if passed non object config */
```
[the remark this answers is not in the document]
Comment (fshovchko) [reply src/modules/esl-carousel/plugin/esl-carousel.plugin.ts:9]: Oh, it's synonym to mimic, nvm

**Review by fshovchko — APPROVED** [review]

File: src/modules/esl-carousel/plugin/autoplay/esl-carousel.autoplay.mixin.ts [hunk 4]
```diff
@@ -1,29 +1,33 @@
 import {ExportNs} from '../../../esl-utils/environment/export-ns';
-import {attr, bind, listen, ready} from '../../../esl-utils/decorators';
+import {bind, listen, ready} from '../../../esl-utils/decorators';
 
 import {ESLCarouselPlugin} from '../esl-carousel.plugin';
 import {ESLCarouselSlideEvent} from '../../core/esl-carousel.events';
 
+export interface ESLCarouselAutoplayConfig {
+  /** Timeout to send next command to the host carousel */
```
Comment (dshovchko) [inline src/modules/esl-carousel/plugin/autoplay/esl-carousel.autoplay.mixin.ts:8]: Is it really timeout? Or delay?

**Review by dshovchko — APPROVED** [review]

File: src/modules/esl-carousel/plugin/autoplay/esl-carousel.autoplay.mixin.ts (hunk 4 above)
Comment (ala-n) [reply src/modules/esl-carousel/plugin/autoplay/esl-carousel.autoplay.mixin.ts:8]: It sounds like we need to discuss variable name as well.
I'll add a note to the cleanup ticket. Let's revisit it separately

**Pull request merged**